Provide correct line info in exceptions.

grokys opened the pull request:
Previously Portable.Xaml wasn't always reporting the correct line/column info when throwing exceptions. This PR adds unit tests for these cases and fixes them; it also correctly unwraps the inner `TargetInvocationException` when a setter throws.

grokys committed:
Fix line/column info for properties.

grokys committed:
Fix line/column/inner exception info for setter that throws.

grokys committed:
Add line info to thrown XamlObjectWriterExceptions.

grokys commented:
The tests are currently failing (but passing when you debug them) due to this code: https://github.com/dotnet/wpf/blob/2cbb1ad9759c32dc575c7537057a29ee7da2e1b2/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/InfosetObjects/XamlReaderSettings.cs#L48

```csharp
private void InitializeProvideLineInfo()
{
	//By default, _provideLineInfo is true if the currently executing process is running in debug mode
	if (Debugger.IsAttached)
	{
		ProvideLineInfo = true;
	}
	else
	{
		ProvideLineInfo = false;
	}
}
```

I assume we don't want to reproduce that behavior. Sigh.

grokys committed:
Make sure tests are run with line info.

XAML engine switches on line info when a debugger is attached, so tests were passing when debugged but failing when ran normally. Sigh.

grokys commented:
Made the tests that rely on line info always provide it.

cwensley commented:
Hey @grokys, this is great, thanks for the update (and the tests are wonderful)!   

I don't see why we can't use the same behaviour of auto enabling line info when debugging, and it appears [the API](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debugger.isattached?view=netframework-4.7.2) exists even in .NET Standard 1.0

grokys commented:
Thanks @cwensley - the reason I didn't implement the same behaviour is that it's confusing that the behaviour changes when debugging! It certainly had me scratching my head for a while.

Is there any good reason to disable line info unless the debugger is attached? Is it a performance hit?

cm4ker commented:
As I see currently we are not using ```ProvideLineInfo``` property from settings that pass to the ```XamlXmlReader``` constructor. (Correct me if I'm wrong here)

But then we start use it this start works: https://github.com/cwensley/Portable.Xaml/blob/ab5526173722b8988bc5ca3c03c8752ce89c0975/src/Portable.Xaml/Portable.Xaml/XamlServices.cs#L128

Not benchmark it, but we are add extra work on every readed node with enabled ```ProvideLineInfo``` property.

cwensley commented:
@grokys, since `ProvideLineInfo` is false by default, there would only be a performance hit when debugging which I would assume would be totally acceptable.

This would ensure line info is provided while debugging unless you set it explicitly (which not many people do).  Since we're not setting this to true by default otherwise, it would actually help users to have it on when debugging imo.

grokys commented:
Sorry, I've been a bit busy and let this slip.

~~The problem with only enabling line info when the debugger is attached is it makes it hard to unit test...~~ no, ignore that - there will be a setting to enable it. Ok, i'll implement it the same as System.Xaml.

grokys committed:
Set ProvideLineInfo if running under debugger.

This matches System.Xaml's behavior: https://github.com/dotnet/wpf/blob/2cbb1ad9759c32dc575c7537057a29ee7da2e1b2/src/Microsoft.DotNet.Wpf/src/System.Xaml/System/Xaml/InfosetObjects/XamlReaderSettings.cs#L48

grokys committed:
Use WithLineInfo to set exception line info.

Rather than passing it in via the ctor. A similar pattern is used in System.Xaml. Also use the same exception messages as System.Xaml.

grokys committed:
Add failing tests for duplicate member exceptions.

grokys committed:
Set line info on duplicate member exceptions.

grokys commented:
@cwensley pushed that change, along with more tests/fixes for line number info for duplicate members.

grokys committed:
Use correct overload here too.

grokys commented:
Actually, hold off on merging this for a little while as I keep finding more places where the exceptions can be improved.

grokys commented:
No, this is ready for review. The other problem I'm seeing is elsewhere and would make more sense as a separate PR. Sorry for the noise!

cm4ker commented:
LGTM, thanks @grokys!
